pkg/services/ngalert/schedule/recording_rule_test.go

```diff
@@ -156,64 +162,146 @@ func TestRecordingRule_Integration(t *testing.T) {
+				grafana_alerting_rule_evaluation_duration_seconds_bucket{org="%[1]d",le="60"} 1
+				grafana_alerting_rule_evaluation_duration_seconds_bucket{org="%[1]d",le="120"} 1
+				grafana_alerting_rule_evaluation_duration_seconds_bucket{org="%[1]d",le="180"} 1
+				grafana_alerting_rule_evaluation_duration_seconds_bucket{org="%[1]d",le="240"} 1
+				grafana_alerting_rule_evaluation_duration_seconds_bucket{org="%[1]d",le="300"} 1
+				grafana_alerting_rule_evaluation_duration_seconds_bucket{org="%[1]d",le="+Inf"} 1
+				grafana_alerting_rule_evaluation_duration_seconds_sum{org="%[1]d"} 0
+				grafana_alerting_rule_evaluation_duration_seconds_count{org="%[1]d"} 1
+				# HELP grafana_alerting_rule_evaluations_total The total number of rule evaluations.
+				# TYPE grafana_alerting_rule_evaluations_total counter
+				grafana_alerting_rule_evaluations_total{org="%[1]d"} 1
+				# HELP grafana_alerting_rule_evaluation_attempts_total The total number of rule evaluation attempts.
+				 # TYPE grafana_alerting_rule_evaluation_attempts_total counter
+				grafana_alerting_rule_evaluation_attempts_total{org="%[1]d"} 1
+				`,
+				rule.OrgID,
+			)
+
+			err := testutil.GatherAndCompare(reg, bytes.NewBufferString(expectedMetric),
+				"grafana_alerting_rule_evaluation_duration_seconds",
+				"grafana_alerting_rule_evaluations_total",
+				"grafana_alerting_rule_evaluation_attempts_total",
+			)
+			require.NoError(t, err)
+		})
+
+		t.Run("reports failure evaluation metrics", func(t *testing.T) {
+			expectedMetric := fmt.Sprintf(
+				`
+				# HELP grafana_alerting_rule_evaluation_failures_total The total number of rule evaluation failures.
+				# TYPE grafana_alerting_rule_evaluation_failures_total counter
+				grafana_alerting_rule_evaluation_failures_total{org="%[1]d"} 1
+				# HELP grafana_alerting_rule_evaluation_attempt_failures_total The total number of rule evaluation attempt failures.
+				# TYPE grafana_alerting_rule_evaluation_attempt_failures_total counter
+				grafana_alerting_rule_evaluation_attempt_failures_total{org="%[1]d"} 1
+				`,
+				rule.OrgID,
+			)
+
+			err := testutil.GatherAndCompare(reg, bytes.NewBufferString(expectedMetric),
+				"grafana_alerting_rule_evaluation_failures_total",
+				"grafana_alerting_rule_evaluation_attempt_failures_total",
+			)
+			require.NoError(t, err)
+		})
+
+		t.Run("no write was performed", func(t *testing.T) {
+			require.Zero(t, writeTarget.RequestsCount)
+		})
 	})
 }
+
+func withQueryForHealth(t *testing.T, health string) models.AlertRuleMutator {
+	t.Helper()
+
+	var expression string
+	switch health {
+	case "ok":
+		expression = `{
+			"datasourceUid": "__expr__",
+			"type":"math",
+			"expression":"2 + 1"
+		}`
+	case "error":
+		expression = `{
+			"datasourceUid": "__expr__",
+			"type":"math",
+			"expression":"$NOTEXIST"
+		}`
+	default:
+		require.Fail(t, "Query generation for health %s is not supported yet", health)
```

Comment:
I wonder if it is just good enough to panic? or even just empty string, which hopefully will cause tests to fail :) having testing.T in mutator function seems strange.

Reply:
Yeah agreed, i like that more :+1: